modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/StatisticsGatherer.java

```diff
@@ -89,7 +102,25 @@ public LocalStatisticsGatheringContext aggregateStatisticsAsync(
         LocalStatisticsGatheringContext inProgressCtx = gatheringInProgress.putIfAbsent(key, ctx);
 
         if (inProgressCtx == null) {
-            CompletableFuture<ObjectStatisticsImpl> f = CompletableFuture.supplyAsync(aggregate, gatherPool);
+            CompletableFuture<ObjectStatisticsImpl> f = CompletableFuture.supplyAsync(
+                () -> {
+                    if (!gatheringLock.enterBusy() || !active) {
+                        if (log.isDebugEnabled())
+                            log.debug("Can't aggregate statistics by key " + key + " due to inactive state.");
+                        gatheringInProgress.remove(key, ctx);
```

Comment:
'active' flag should be changed under busy lock.
If you failed to obtain lock, then someone who holds it must cancel gathering.

Otherwise,  without holding the lock you will remove key under race, don't you?

Reply:
Once, obtaing the lock you will never release it if active==false.

Reply:
Right, fixed.